### ISSUE TITLE
WebSocket support

### ISSUE THREAD

**Issue by eye-dee** [vukoye/xmpp_dart#67]
Hello,
As far as I can see the library doesn't have websocket support
Are there any plans to do so?

**Comment by TatankaConCube** [comment]
it already has support for WebSocket, try version [0.4.2-dev.1)](https://pub.dev/packages/xmpp_stone/versions/0.4.2-dev.1), it works well in our project

**Comment by eye-dee** [comment]
Thank you, works fine!

The only thing I needed to change is a possibility to specify url path for web socket connection:

class XmppWebSocketHtml: 
```

 _socket = WebSocketChannel.connect(
      Uri(
        scheme: 'wss',
        host: host,
        port: port,
        path: 'xmpp-websocket' <-------------- Path here
      ),
      protocols: ['xmpp'],
    );
```

**Comment by eldhomt** [comment]
@sausageRoll I tried your suggestion. It is giving the following error:
```
[log] ---Xmpp Receiving:---
[log] <stream:error xmlns:stream="http://etherx.jabber.o…ietf:params:xml:ns:xmpp-streams"/></stream:error>
[log] ---Xmpp Receiving:---
[log] <close xmlns='urn:ietf:params:xml:ns:xmpp-framing'/>
[log] D/[Connection]: Handle connection done
[log] D/[Connection]: State: XmppConnectionState.ForcefullyClosed
[log] D/[ReconnectionManager]: Connection forcefully closed!
[log] D/[ReconnectionManager]: TimeOut is: 2000 reconnection counter 0
```

**Comment by eye-dee** [comment]
@eldhomt The path is different for different servers. Most of the time users should be able to connect without it. But we have a specific deployment and it requires path in URI:
```
wss://host:port/xmpp-websocket
```

If that's what you tried
My suggestion would be exposing that path as a configuration parameter similar to `host` and `port`

**Comment by eldhomt** [comment]
@sausageRoll Yes, I have a used a different path. In my case the schema is `ws `and path is `ws://host:port/ws` and I am using openfire server. I have tested the same with React web application and is working fine.

**Comment by eye-dee** [comment]
@eldhomt Make sure you use WebSocket implementation, because there are two different implementations for sockets.

```
import 'connection/XmppWebsocketApi.dart'
  if (dart.library.io) 'connection/XmppWebsocketIo.dart'
  if (dart.library.html) 'connection/XmppWebsocketHtml.dart' as xmppSocket;
```

I had to change it to:
```
import 'connection/XmppWebsocketHtml.dart' as xmppSocket;
```

**Comment by eldhomt** [comment]
@sausageRoll No Luck. I am getting the same error. Few others have reported the same issue at https://github.com/vukoye/xmpp_dart/issues/71.

**Comment by szotp** [comment]
According to the [spec](https://datatracker.ietf.org/doc/html/rfc7395#section-3.4) communication with web socket should start with something like this
```xml
<open xmlns="urn:ietf:params:xml:ns:xmpp-framing" to="example.com" version="1.0" />
```
but nothing like that exists in this repository. 

Sad to be saying this, but this library is completely unusable in its current state (at least if you want web sockets).
1. The example code is wrong or obsolete.
2. There are no version tags.
3. [This class](https://github.com/vukoye/xmpp_dart/blob/null_safety/lib/src/connection/XmppWebsocketIo.dart) makes no sense, it's regular socket pretenting to be a web socket. 
4. You can't connect with Url.
5. Web socket doesn't even connect properly

**Comment by codesxt** [comment]
@eldhomt Have you been able to connect? I still can't make it work even with the proposed solutions :(

**Comment by codesxt** [comment]
I could not make it work by just changing the URI parsing process as the first comments suggested. When I checked ejabberd logs, the server was receiving my requests but throwing this error:

```
2022-04-25 10:47:32.143 [error] <0.2316.0>@xmpp_socket:parse:422 CRASH REPORT Process <0.2316.0> with 0 neighbours exited with reason: bad argument in call to fxml_stream:parse(undefined, <<"<?xml version='1.0'?><stream:stream xmlns='jabber:client' version='1.0' xmlns:stream='http://et...">>) in xmpp_socket:parse/2 line 422
```

So I thought that the error should be somewhere else and checked @szotp 's answer. I checked [RFC-3975 section 3.4](https://datatracker.ietf.org/doc/html/rfc7395#section-3.4) and changed the `streamOpeningString` variable and it worked, instead of the stream:stream opening string that works on regular sockets.

So maybe there should be a condition to select the proper stream opening stream according to the specific socket type. I wrote this condition to make it work for both sockets:

```dart
void _openStream() {
    var streamOpeningString =
        (_socket.runtimeType.toString() == 'XmppWebSocketHtml')
            ? """
<open xmlns="urn:ietf:params:xml:ns:xmpp-framing"
             to='${fullJid.domain}'
             version="1.0" />
"""
            : """
<?xml version='1.0'?>
<stream:stream xmlns='jabber:client' version='1.0' xmlns:stream='http://etherx.jabber.org/streams'
to='${fullJid.domain}'
xml:lang='en'
>
""";
    write(streamOpeningString);
  }
```

**Comment by eldhomt** [comment]
@codesxt Thanks. That fixes the issue. Working fine with all the platforms now.

**Comment by TatankaConCube** [comment]
today we released a new version [0.4.4-dev.1](https://pub.dev/packages/xmpp_stone/versions/0.4.4-dev.1) with fixes and requirements from this thread

**Comment by HaoCherHong** [comment]
> today we released a new version [0.4.4-dev.1](https://pub.dev/packages/xmpp_stone/versions/0.4.4-dev.1) with fixes and requirements from this thread

@TatankaConCube Does this branch supports web?

**Comment by TatankaConCube** [comment]
yes, sure, but the latest version, for now, is 0.4.4-dev.3
